### ISSUE TITLE
Can validation set be used as unlabel data during training?

### ISSUE THREAD

**Issue by HeimingX** [vikasverma1077/ICT#4]
Thanks for your inspiring idea and accompanying code.

I found that validation set is also used as one part of unlabel data during training your model. I wonder if this approach is reasonable? If the answer is yes, can the test set also be added to the unlabel set for training, since more unlabel data will further imporve the performance.

Looking forward to your reply!
Thanks

**Comment by vikasverma1077** [comment]
yes, validation data can be used as unlabeled data (this has been done in the previous papers as well). We dont use the test data as unlabeled data since we assume that we dont have access to test data while training.

**Comment by HeimingX** [comment]
Thanks for your reply.
I have read some semi-supervised learning papers[1,2,3] and their published source code again. But I found that none of them use the validation set during the training phase. Specifically, the file "mean-teacher/pytorch/data-local/bin/link_cifar10_train.sh" in mean-teacher method's source code only contains 45k images as unlabelled data for training. For VAT method, the "prepare_dataset" function in file ''vat_tf/cifar10.py" shows that the unlabelled data also do not contain any validation data. For the paper[3], the file "realistic-ssl-evaluation/build_tfrecords.py" is used to split the dataset and the training data does not contain validation data too. 
So, I wonder which paper you refer to use the validation set as one part of unlabelled data during training.
(PS: personally, the selection of the final model depends on the performance of the validation set in many semi-supervised methods and the use of validation set information in training phase is not very reasonable. )

Looking forward to your reply!
Thanks again

[1]Mean teachers are better role models: Weight-averaged consistency targets improve semi-supervised deep learning results
[2]Virtual Adversarial Training: A Regularization Method for Supervised and Semi-Supervised Learning
[3]Realistic Evaluation of Deep Semi-Supervised Learning Algorithms

**Comment by vikasverma1077** [comment]
"PS: personally, the selection of the final model depends on the performance of the validation set in many semi-supervised methods and the use of validation set information in training phase is not very reasonable."  We are choosing the best model on the validation set and report the test accuracy on that model. If we had used the validation set along with its labels in the training, then the model would have overfitted to this validation data. Consequently, the resulting best model on this validation data would not have generalized well on the test data. However, since we dont use any class labels from the validation data, there is no such overfitting. Does this sound reasonable to you now?

**Comment by vikasverma1077** [comment]
On your question about whether the test data can be used as unlabeled data: it can be if you assume that you have access to that. And if you apply consistency loss to test data as well, it will closely relate to Transductive Learning.
